protocols/request-response/src/lib.rs

```diff
@@ -0,0 +1,583 @@
+};
+
+/// An inbound request or response.
+#[derive(Debug)]
+pub enum RequestResponseMessage<TRequest, TResponse> {
+    /// A request message.
+    Request {
+        /// The request message.
+        request: TRequest,
+        /// The sender of the request who is awaiting a response.
+        ///
+        /// See [`RequestResponse::send_response`].
+        channel: ResponseChannel<TResponse>,
+    },
+    /// A response message.
+    Response {
+        /// The ID of the request that produced this response.
+        ///
+        /// See [`RequestResponse::send_request`].
+        request_id: RequestId,
+        /// The response message.
+        response: TResponse
+    },
+}
+
+/// The events emitted by a [`RequestResponse`] protocol.
+#[derive(Debug)]
+pub enum RequestResponseEvent<TRequest, TResponse> {
+    /// An incoming message (request or response).
+    Message {
+        /// The peer who sent the message.
+        peer: PeerId,
+        /// The incoming message.
+        message: RequestResponseMessage<TRequest, TResponse>
+    },
+    /// An outbound request failed.
+    OutboundFailure {
+        /// The peer to whom the request was sent.
+        peer: PeerId,
+        /// The (local) ID of the failed request.
+        request_id: RequestId,
+        /// The error that occurred.
+        error: OutboundFailure,
+    },
+    /// An inbound request failed.
+    InboundFailure {
+        /// The peer from whom the request was received.
+        peer: PeerId,
+        /// The error that occurred.
+        error: InboundFailure,
+    },
+}
+
+/// Possible failures occurring in the context of sending
+/// an outbound request and receiving the response.
+#[derive(Debug)]
+pub enum OutboundFailure {
+    /// The request could not be sent because a dialing attempt failed.
+    DialFailure,
+    /// The request timed out before a response was received.
+    ///
+    /// It is not known whether the request may have been
+    /// received (and processed) by the remote peer.
+    Timeout,
+    /// The connection closed before a response was received.
+    ///
+    /// It is not known whether the request may have been
+    /// received (and processed) by the remote peer.
+    ConnectionClosed,
+    /// The remote supports none of the requested protocols.
+    UnsupportedProtocols,
+}
+
+/// Possible failures occurring in the context of receiving an
+/// inbound request and sending a response.
+#[derive(Debug)]
+pub enum InboundFailure {
+    /// The inbound request timed out, either while reading the
+    /// incoming request or before a response is sent, i.e. if
+    /// [`RequestResponse::send_response`] is not called in a
+    /// timely manner.
+    Timeout,
+    /// The local peer supports none of the requested protocols.
+    UnsupportedProtocols,
+}
+
+/// A channel for sending a response to an inbound request.
+///
+/// See [`RequestResponse::send_response`].
+#[derive(Debug)]
+pub struct ResponseChannel<TResponse> {
```

Comment:
Other similar suggestion: maybe add a `pub fn is_known_stale(&self) -> bool` method to `ResponseChannel`?

In protocols where building the response to a request is CPU-intensive or IO-intensive, it might be a good idea to call `is_known_stale()` beforehand and discard the `ResponseChannel` is `true` is returned.

Additionally, in situations where the program accumulates incoming requests at faster rate than it processes them, it would help to discard `ResponseChannel`s that we know already have timed out. Otherwise, the program would spend most of its time building responses to timed out requests.

Reply:
Sounds good. I added `ResponseChannel::is_open` in https://github.com/libp2p/rust-libp2p/pull/1596/commits/7bfdff5c630e202f4dacf18e42c0a9e5dd18e822.